### DISCUSSION TITLE
Memory used by MUMPS is not being cleaned?

### DISCUSSION THREAD

**Discussion by miguelcoolchips** [firedrakeproject/firedrake#2925] [General]
I ran the following example with [memray](https://github.com/bloomberg/memray) (python -m memray run --native test_multiple_solvers.py) 
```python
import firedrake as fd
import petsc4py
import gc


mesh = fd.UnitSquareMesh(1000, 1000)
V = fd.FunctionSpace(mesh, "CG", 1)
u = fd.Function(V)
v = fd.TestFunction(V)
a = fd.inner(fd.grad(u), fd.grad(v)) * fd.dx + u * v * fd.dx
f = fd.Constant(1.0, domain=mesh)
L = f * v * fd.dx
bc = fd.DirichletBC(V, 0.0, "on_boundary")
problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver = fd.NonlinearVariationalSolver(problem)

solver.solve()
print("Solved once")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)

problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver2 = fd.NonlinearVariationalSolver(problem)

solver2.solve()
print("Solved twice")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)

problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver3 = fd.NonlinearVariationalSolver(problem)

solver3.solve()
print("Solved thrice")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)

problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver4 = fd.NonlinearVariationalSolver(problem)

solver4.solve()
print("Solved four times")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)

problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver5 = fd.NonlinearVariationalSolver(problem)

solver5.solve()
print("Solved five times")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)

problem = fd.NonlinearVariationalProblem(a - L, u, bcs=bc)
solver6 = fd.NonlinearVariationalSolver(problem)

solver6.solve()
print("Solved six times")
gc.collect()
petsc4py.PETSc.garbage_cleanup(mesh._comm)
petsc4py.PETSc.garbage_cleanup(mesh.comm)
```
and the flamegraph looks as follows:
<img width="1911" alt="image" src="https://github.com/firedrakeproject/firedrake/assets/115098306/0952718d-21d7-4602-a95e-84c86523dff3">
The heap memory goes all the way up to 22 GB (orange line at the top). Most of the memory is consumed by MUMPS, but it seems that is not cleaned? If I delete each solver right after using it (`del solver`), the flamegraph looks cleaner and the heap memory only reaches 8 GB.
<img width="1913" alt="image" src="https://github.com/firedrakeproject/firedrake/assets/115098306/fd95e33d-69a1-4db0-a8fb-5ebaa99fe7ef">

The calls to `gc.collect()` and the PETSc garbage collector did not make much difference. Is the decision to leave the MUMPS memory untouched intentional? Perhaps it's to allow its reuse for multiple calls to the solver.

**Comment by wence-** [comment]
That is intentional. None of the solvers ever go out of scope, so they should not be collected. Basically a solver holds on to all its state for reuse (so in the mumps case, exactly so that the factorisation can be reused if you solve again with the same solver).

**Reply by miguelcoolchips** [reply]
Is there a way to reset the solver state without deleting it?

**Reply by wence-** [reply]
`solver.snes.reset()` might do it, but it's possible that we don't correctly deal with the teardown everywhere. How much state do you want to destroy? Just the PC setup? Also the allocated matrices?

**Reply by miguelcoolchips** [reply]
Just the PC setup which seems to occupy the most memory. I tried doing `solver.snes.ksp.pc.reset()`, and it did the trick. Thanks!
